### PR TITLE
Add exclude-class parameter 

### PR DESCRIPTION
**Pull request by UshakovN**

### CODE REVIEW

File: suggest_trie.go [hunk 1]
```diff
@@ -70,12 +71,23 @@ type SuggestTrieBuilder struct {
...
+      }
+    }
+  }
   class := ""
   if knownClass, ok := item.OriginalItem.Data["class"]; ok {
     class = strings.ToLower(knownClass.(string))
   }
   for _, suggest := range s.Suggest {
-    if suggest.Class == class {
+    if suggest.Class == class && Equal(suggest.Classes, classes) {
```
Comment (ashagraev) [inline suggest_trie.go:90]: I think that's not what we mean here. We mean: that if there's at least one class from the list, the item should remain in the answer. So they should work like "or", not "and".

File: suggest_trie.go [hunk 2]
```diff
@@ -70,12 +71,23 @@ type SuggestTrieBuilder struct {
 }
 
 func (s *SuggestTrieBuilder) addItem(maxItemsPerPrefix int, item *SuggestTrieItem) {
+  var classes []string
+  if knownClasses, ok := item.OriginalItem.Data["classes"]; ok {
```
Comment (ashagraev) [inline suggest_trie.go:75]: Why do we need this complexity here? Shouldn't we just make all these classes strings?

File: common.go [hunk 3]
```diff
@@ -11,3 +11,16 @@ func PrepareCheckMap(values []string) map[string]bool {
   }
   return valuesMap
 }
+
+func AtLeastOneEqual(first, second []string) bool {
```
Comment (ashagraev) [inline common.go:15]: We only use this function once. Why complicate things with generics?

Comment (UshakovN) [reply common.go:15]: first, second []string

File: common.go [hunk 4]
```diff
@@ -11,3 +11,16 @@ func PrepareCheckMap(values []string) map[string]bool {
...
+  if len(first) > len(second) {
+    first, second = second, first
+  }
+  for idx, valueFirst := range first {
```
Comment (ashagraev) [inline common.go:19]: I don't get how this is supposed to work. If the idea is to compare two arrays as sets (meaning, two sets have a non-empty intersection), that's clearly the wrong algorithm. If the idea is to see if there's a position in which two arrays have a similar item, I don't understand how this relates to the original task.

File: suggest_trie.go [hunk 5]
```diff
@@ -70,12 +72,17 @@ type SuggestTrieBuilder struct {
...
+    if err := json.Unmarshal(knownClasses, &classes); err == nil {
+    }
+  }
   class := ""
   if knownClass, ok := item.OriginalItem.Data["class"]; ok {
     class = strings.ToLower(knownClass.(string))
   }
   for _, suggest := range s.Suggest {
-    if suggest.Class == class {
+    if suggest.Class == class || AtLeastOneEqual(suggest.Classes, classes) {
```
Comment (ashagraev) [inline suggest_trie.go:85]: I think we can simplify things by just adding the Class to the Classes array during the data load. After that, we can process only the Classes array without even noticing that there's a Class field. Yet this will keep the implementation backward-compatible.

File: suggest_trie.go [hunk 6]
```diff
@@ -70,12 +72,17 @@ type SuggestTrieBuilder struct {
 }
 
 func (s *SuggestTrieBuilder) addItem(maxItemsPerPrefix int, item *SuggestTrieItem) {
+  var classes []string
+  if knownClasses, err := json.Marshal(item.OriginalItem.Data["classes"]); err == nil {
```
Comment (ashagraev) [inline suggest_trie.go:76]: 1. Do not put conditions in the form err == nil. It's not a go way :) The idiomatic way to write code is to check if err != nil, so when you're doing the opposite, you confuse the reader.
2. I don't see any error handing here.
3. In fact, I imagine why it's not convenient to handle an error right there — because the function is no return. But it's not a good idea to just drop errors. Instead, just make the needed transformation at the data load. Make an additional field in the OriginalItem (or even the embracing SuggestTrieItem?) to avoid data transformation inside the business logic.

File: suggest.go [hunk 7]
```diff
@@ -51,7 +51,8 @@ func (pt *ProtoTransformer) TransformTrie(builder *SuggestTrieBuilder) (*stpb.Su
   }
   for _, suggest := range builder.Suggest {
     trieItems := &stpb.ClassItems{
-      Class: suggest.Class,
+      Class:   suggest.Class,
```
Comment (ashagraev) [inline suggest.go:54]: In fact, we can remove the class field from the stpb.ClassItems, see the comment below. That would simplify things a bit.

File: go.mod [hunk 8]
```diff
@@ -1,6 +1,6 @@
 module main
 
-go 1.17
+go 1.18
```
Comment (ashagraev) [inline go.mod:3]: If we don’t use generics, let’s keep the go version. Otherwise, I’d like to know why we need to bump it.

File: common.go [hunk 9]
```diff
@@ -2,12 +2,39 @@ package main
...
 
-func PrepareCheckMap(values []string) map[string]bool {
-  valuesMap := map[string]bool{}
-  for _, value := range values {
-    if value != "" {
-      valuesMap[strings.ToLower(value)] = true
+func Contains(source []string, item string) bool {
+  sourceMap := map[string]bool{}
+  for _, value := range source {
+    sourceMap[strings.ToLower(value)] = true
```
Comment (ashagraev) [inline common.go:8]: If you already iterated through all the items, why complicate things? :)

```
if strings.ToLower(value) == strings.ToLower(item) {
  return true
}
```

File: common.go [hunk 10]
```diff
@@ -2,12 +2,39 @@ package main
...
+  }
+  for _, secondValue := range second {
+    if _, ok := firstMap[strings.ToLower(secondValue)]; ok {
+      return true
+    }
+  }
+  return false
+}
+
+func Equal(first, second []string) bool {
```
Comment (ashagraev) [inline common.go:29]: You can go with https://pkg.go.dev/reflect#DeepEqual in this case :)

File: suggest.go [hunk 11]
```diff
@@ -93,16 +94,18 @@ func BuildSuggest(items []*Item, maxItemsPerPrefix int, postfixWeightFactor floa
...
     for i := 1; i < len(parts); i++ {
       builder.Add(0, strings.Join(parts[i:], " "), veroheadItemsCount, &SuggestTrieItem{
         Weight:       item.Weight * postfixWeightFactor,
         OriginalItem: item,
+        Classes:      item.Classes,
       })
     }
     if (idx+1)%100000 == 0 {
-      log.Printf("addedd %d items of %d to suggest", idx+1, len(items))
+      log.Printf("added %d items of %d to suggest", idx+1, len(items))
```
Comment (ashagraev) [inline suggest.go:108]: Fix in a separate PR plz. Let's not mix up meaningful changes and small unrelated fixes.

File: suggest_items.go [hunk 12]
```diff
@@ -31,11 +32,25 @@ func NewItem(line string, policy *bluemonday.Policy) (*Item, error) {
   if err := json.Unmarshal([]byte(parts[2]), &data); err != nil {
     return nil, fmt.Errorf("cannot parse data json: %v", err)
   }
+  classes := make([]string, 0)
```
Comment (ashagraev) [inline suggest_items.go:35]: `var classes []string` is the preferred style of defining empty slices: https://github.com/zchee/golang-wiki/blob/master/CodeReviewComments.md#declaring-empty-slices

File: suggest_items.go [hunk 13]
```diff
@@ -31,11 +32,25 @@ func NewItem(line string, policy *bluemonday.Policy) (*Item, error) {
...
+  dataClasses, err := json.Marshal(data["classes"])
+  if err != nil {
+    return nil, fmt.Errorf("cannot convert data in json: %v", err)
```
Comment (ashagraev) [inline suggest_items.go:38]: in => to

File: suggest_items.go [hunk 14]
```diff
@@ -31,11 +32,25 @@ func NewItem(line string, policy *bluemonday.Policy) (*Item, error) {
...
+  }
+  if err := json.Unmarshal(dataClasses, &classes); err != nil {
+    return nil, fmt.Errorf("cannot parse data json: %v", err)
+  }
+  if deprecatedClass, ok := data["class"].(string); ok {
+    if !Contains(classes, deprecatedClass) {
```
Comment (ashagraev) [inline suggest_items.go:44]: I’m not sure if we should make this check here. Duplicates in this array do not impact the results in any case. So it's up to user, basically.
